src/raftstore/store/store.rs

```diff
@@ -273,13 +273,62 @@ impl<T: Transport, C: PdClient> Store<T, C> {
+
+                // If this peer detects the leader is missing for a long long time,
+                // it should consider itself as a stale peer which is removed from
+                // the original cluster.
+                // This most likely happens in the following scenario:
+                // 1. At first, there are three node A, B, C in the cluster, and A is leader.
+                // Node B gets down. And then A adds D, E, F into the cluster.
+                // Node D becomes leader of the new cluster, and then removes node A, B, C.
+                // After all these node in and out, now the cluster has node D, E, F.
+                // If node B goes up at this moment, it still thinks it is one of the cluster
+                // and has peers A, C. However, it could not reach A, C since they are removed
+                // from the cluster or probably destroyed.
+                // Meantime, D, E, F would not reach B, since it's not in the cluster anymore.
+                // 2. A peer, B is initialized as a replicated peer without data after
+                // receiving a single raft AE message. But then it goes through some process like 1,
+                // it's removed out of the region and wouldn't be contacted anymore.
+                // In both cases, Node B would notice that the leader is missing for a long time,
+                // and it would check with pd to confirm whether it's still a member of the cluster.
+                // If not, it destroys itself as a stale peer which is removed out already.
+                let duration = peer.since_leader_missing();
+                if duration >= self.cfg.max_leader_missing_duration {
+                    info!("{} detects leader missing for a long time. To check with pd whether \
+                           it's still valid",
+                          peer.tag);
+                    // reset the leader missing time to avoid sending the same tasks to
+                    // PD worker continuously on everytime raft ready event triggers
+                    peer.reset_leader_missing_time();
+                    if peer.is_initialized() {
+                        // for peer B in case 1 above
+                        let task = PdTask::ValidatePeer {
+                            peer: peer.peer.clone(),
+                            region: peer.region().clone(),
+                        };
+                        if let Err(e) = self.pd_worker.schedule(task) {
+                            error!("{} failed to notify pd: {}", peer.tag, e)
+                        }
+                    } else {
+                        // for peer B in case 2 above
+                        // directly destroy peer without data since it doesn't have region range,
+                        // so that it doesn't have the correct region start_key to validate peer with PD
+                        region_to_be_destroyed.push((region_id, peer.peer.clone()))
```

Comment:
A test case is added for the gc of uninitialized peer.
The region_id is cached for direct call to `destroy_peer()` without reference to the peer. I think we should keep it this way.